harness/determined/pytorch/_pytorch_context.py

```diff
@@ -57,4 +82,224 @@ def get_lr_scheduler(self) -> Optional[pytorch.LRScheduler]:
+            if self.hvd_config.use:
+                check.eq(
+                    self.hvd_config.aggregation_frequency,
+                    1,
+                    "Mixed precision training (AMP) is not supported with "
+                    "aggregation frequency > 1.",
+                )
+
+            check.true(
+                torch.cuda.is_available(),
+                "Mixed precision training (AMP) is supported only on GPU slots.",
+            )
+            check.false(
+                not self.hvd_config.use and self.n_gpus > 1,
+                "To enable mixed precision training (AMP) for parallel training, "
+                'please set `resources["optimized_parallel"] = True`.',
+            )
+
+            logging.info(
+                f"Enabling mixed precision training with opt_level: {self.mixed_precision}."
+            )
+            model, optimizers = apex.amp.initialize(
+                model,
+                optimizers,
+                opt_level=self.mixed_precision,
+                verbosity=1
+                if self.distributed.get_rank() == 0 or self.env.experiment_config.debug_enabled()
+                else 0,
+            )
+            self.model = model
+            if not isinstance(optimizers, list):
+                self.optimizers = [optimizers]
+        return model, optimizers
+
+    def _should_communicate_and_update(self) -> bool:
+        if self.current_batch_idx is None:
+            raise det.errors.InternalException("Training hasn't started.")
+        return (self.current_batch_idx + 1) % self.hvd_config.aggregation_frequency == 0
+
+    def _backward(self, loss: torch.Tensor):  # type: ignore
+        if self._use_amp():
+            with apex.amp.scale_loss(loss, self.optimizers) as scaled_loss:
+                scaled_loss.backward()
+
+                if self.hvd_config.use and self._should_communicate_and_update():
+                    # When using horovod, we need to finish communicating gradient
+                    # updates before they are unscaled which happens when we exit
+                    # of this context manager.
+                    for optimizer in self.optimizers:
+                        optimizer.synchronize()
+        else:
+            loss.backward()
+
+            # Communication needs to be synchronized so that is completed
+            # before we apply gradient clipping and `step()`.
+            if self.hvd_config.use and self._should_communicate_and_update():
+                for optimizer in self.optimizers:
+                    optimizer.synchronize()
+
+    def _average_gradients(self, parameters: Any, divisor: int) -> None:
+        check.gt_eq(divisor, 1)
+        if divisor == 1:
+            return
+
+        divisor_value = float(divisor)
+        for p in filter(lambda param: param.grad is not None, parameters):
+            p.grad.data.div_(divisor_value)
+
+    def _step_optimizer(self, optimizer: torch.optim.Optimizer) -> None:  # type: ignore
```

Comment:
question: we had previously talked about handing the optimizer a closure to execute for gradient clipping, is that outside the scope of this PR?

Reply:
Sorry forgot that. What do you mean by `closure`, passing in a function?

Reply:
Added an argument to pass in a function that accepts the parameters as an argument and backward compatibility for `on_before_optimizer_step` callback.